File: src/providers/google.ts
import { ProviderError, type ChatEvent, type ChatRequest, type ProviderAdapter, type ProviderCapabilities, type Usage } from "./types.js";

interface GoogleOptions { apiKey: string; baseUrl?: string; fetchImpl?: typeof fetch; }

export class GoogleAdapter implements ProviderAdapter {
  readonly id = "google";
  private readonly apiKey: string; private readonly baseUrl: string; private readonly fetchImpl: typeof fetch;
  constructor(options: GoogleOptions) { if (!options.apiKey) throw new Error("Google requires GOOGLE_GENERATIVE_AI_API_KEY."); this.apiKey = options.apiKey; this.baseUrl = (options.baseUrl ?? "https://generativelanguage.googleapis.com").replace(/\/+$/, ""); this.fetchImpl = options.fetchImpl ?? fetch; }
  capabilities(): ProviderCapabilities { return { streaming: true, tools: true, usage: true, cancellation: true, jsonMode: true }; }
  async *stream(request: ChatRequest): AsyncIterable<ChatEvent> {
    yield { type: "start", provider: this.id, model: request.model };
    const system = request.messages.filter((message) => message.role === "system").map((message) => message.content).join("\n\n");
    const contents = request.messages.filter((message) => message.role !== "system").map((message) => ({ role: message.role === "assistant" ? "model" : "user", parts: [{ text: message.content }] }));
    const body = { contents, ...(system ? { systemInstruction: { parts: [{ text: system }] } } : {}), generationConfig: { maxOutputTokens: request.maxOutputTokens, ...(request.temperature === undefined ? {} : { temperature: request.temperature }) } };
    const response = await this.fetchImpl(`${this.baseUrl}/v1beta/models/${encodeURIComponent(request.model)}:streamGenerateContent?alt=sse`, { method: "POST", headers: { "content-type": "application/json", "x-goog-api-key": this.apiKey }, body: JSON.stringify(body), ...(request.signal ? { signal: request.signal } : {}) });
    if (!response.ok) throw new ProviderError(this.id, `Provider returned HTTP ${response.status}: ${(await response.text()).slice(0, 300)}`, { status: response.status, retryable: response.status === 429 || response.status >= 500 });
    if (!response.body) throw new ProviderError(this.id, "Provider returned an empty response body.", { retryable: true });
    const reader = response.body.pipeThrough(new TextDecoderStream()).getReader(); let buffer = ""; let usage: Usage | undefined;
    try {
      while (true) { const { value, done } = await reader.read(); if (done) break; buffer += value; const lines = buffer.split(/\r?\n/); buffer = lines.pop() ?? "";
        for (const line of lines) { if (!line.startsWith("data:")) continue; let data: unknown; try { data = JSON.parse(line.slice(5).trim()) as unknown; } catch { continue; } if (!data || typeof data !== "object") continue;
          const record = data as Record<string, unknown>; const first = (Array.isArray(record.candidates) ? record.candidates[0] : undefined) as Record<string, unknown> | undefined; const content = first?.content as Record<string, unknown> | undefined; const parts = Array.isArray(content?.parts) ? content.parts : []; const text = (parts[0] as Record<string, unknown> | undefined)?.text; if (typeof text === "string") yield { type: "delta", text };
          const metadata = record.usageMetadata as Record<string, unknown> | undefined; if (typeof metadata?.promptTokenCount === "number" && typeof metadata.candidatesTokenCount === "number") usage = { inputTokens: metadata.promptTokenCount, outputTokens: metadata.candidatesTokenCount, totalTokens: typeof metadata.totalTokenCount === "number" ? metadata.totalTokenCount : metadata.promptTokenCount + metadata.candidatesTokenCount, estimated: false };
        }
      }
      if (usage) yield { type: "usage", usage }; yield { type: "done", finishReason: null };
    } finally { reader.releaseLock(); }
  }
}

File: src-rust/providers/google.rs
use async_trait::async_trait;
use futures_util::StreamExt;
use serde::Serialize;

use crate::providers::types::{
    error_event, ChatEvent, ChatMessage, ChatRequest, ChatRole, ProviderAdapter,
    ProviderCapabilities, ProviderError, Usage,
};

#[derive(Debug, Clone)]
pub struct GoogleAdapter {
    pub api_key: String,
    pub base_url: String,
}

impl GoogleAdapter {
    pub fn new(api_key: String, base_url: Option<String>) -> Result<Self, String> {
        if api_key.is_empty() {
            return Err("Google requires GOOGLE_GENERATIVE_AI_API_KEY.".to_string());
        }
        Ok(Self {
            api_key,
            base_url: base_url
                .unwrap_or_else(|| "https://generativelanguage.googleapis.com".to_string())
                .trim_end_matches('/')
                .to_string(),
        })
    }
}

#[derive(Serialize)]
struct GooglePart {
    text: String,
}

#[derive(Serialize)]
struct GoogleContent {
    role: &'static str,
    parts: Vec<GooglePart>,
}

#[derive(Serialize)]
struct GoogleGenerationConfig {
    max_output_tokens: usize,
    #[serde(skip_serializing_if = "Option::is_none")]
    temperature: Option<f64>,
}

#[derive(Serialize)]
struct GoogleBody {
    contents: Vec<GoogleContent>,
    #[serde(skip_serializing_if = "Option::is_none")]
    system_instruction: Option<serde_json::Value>,
    generation_config: GoogleGenerationConfig,
}

#[async_trait]
impl ProviderAdapter for GoogleAdapter {
    fn id(&self) -> &str {
        "google"
    }

    fn capabilities(&self) -> ProviderCapabilities {
        ProviderCapabilities {
            streaming: true,
            tools: true,
            usage: true,
            cancellation: true,
            json_mode: true,
        }
    }

    async fn stream(
        &self,
        request: &ChatRequest,
        emit: &mut (dyn FnMut(ChatEvent) + Send),
    ) -> Result<(), ProviderError> {
        emit(ChatEvent::Start {
            provider: "google".to_string(),
            model: request.model.clone(),
        });

        let system = request
            .messages
            .iter()
            .filter(|m| m.role == ChatRole::System)
            .map(|m| m.content.as_str())
            .collect::<Vec<_>>()
            .join("\n\n");
        let contents: Vec<GoogleContent> = request
            .messages
            .iter()
            .filter(|m| m.role != ChatRole::System)
            .map(|m| GoogleContent {
                role: if m.role == ChatRole::Assistant { "model" } else { "user" },
                parts: vec![GooglePart {
                    text: m.content.clone(),
                }],
            })
            .collect();

        let body = GoogleBody {
            contents,
            system_instruction: (!system.is_empty()).then(|| {
                serde_json::json!({ "parts": [{ "text": system }] })
            }),
            generation_config: GoogleGenerationConfig {
                max_output_tokens: request.max_output_tokens,
                temperature: request.temperature,
            },
        };

        let encoded_model = urlencoding_light(&request.model);
        let url = format!(
            "{}/v1beta/models/{}:streamGenerateContent?alt=sse",
            self.base_url, encoded_model
        );

        let client = reqwest::Client::new();
        let response = client
            .post(&url)
            .header("content-type", "application/json")
            .header("x-goog-api-key", &self.api_key)
            .json(&body)
            .send()
            .await
            .map_err(|e| ProviderError::new("google", format!("Request failed: {e}"), None, true))?;

        if !response.status().is_success() {
            let status = response.status().as_u16();
            let detail = response.text().await.unwrap_or_default();
            return Err(ProviderError::new(
                "google",
                format!(
                    "Provider returned HTTP {status}: {}",
                    &detail[..detail.len().min(300)]
                ),
                Some(status),
                status == 429 || status >= 500,
            ));
        }

        let mut stream = response.bytes_stream();
        let mut buffer = String::new();
        let mut usage: Option<Usage> = None;

        while let Some(chunk) = stream.next().await {
            let chunk = match chunk {
                Ok(bytes) => bytes,
                Err(e) => {
                    emit(error_event("google", format!("Stream read failed: {e}")));
                    break;
                }
            };
            buffer.push_str(&String::from_utf8_lossy(&chunk));
            let lines: Vec<String> = buffer.split('\n').map(|s| s.to_string()).collect();
            buffer = lines.last().cloned().unwrap_or_default();

            for line in lines[..lines.len().saturating_sub(1)].iter() {
                let Some(payload) = line.trim_start().strip_prefix("data:") else {
                    continue;
                };
                let Ok(data) = serde_json::from_str::<serde_json::Value>(payload.trim()) else {
                    continue;
                };
                if let Some(text) = data
                    .get("candidates")
                    .and_then(|c| c.as_array())
                    .and_then(|c| c.first())
                    .and_then(|c| c.get("content"))
                    .and_then(|c| c.get("parts"))
                    .and_then(|p| p.as_array())
                    .and_then(|p| p.first())
                    .and_then(|p| p.get("text"))
                    .and_then(|t| t.as_str())
                {
                    if !text.is_empty() {
                        emit(ChatEvent::Delta {
                            text: text.to_string(),
                        });
                    }
                }
                if let Some(metadata) = data.get("usageMetadata") {
                    let prompt = metadata.get("promptTokenCount").and_then(|v| v.as_u64());
                    let candidates = metadata.get("candidatesTokenCount").and_then(|v| v.as_u64());
                    if let (Some(prompt), Some(candidates)) = (prompt, candidates) {
                        let total = metadata
                            .get("totalTokenCount")
                            .and_then(|v| v.as_u64())
                            .map(|t| t as usize)
                            .unwrap_or((prompt + candidates) as usize);
                        usage = Some(Usage {
                            input_tokens: prompt as usize,
                            output_tokens: candidates as usize,
                            total_tokens: total,
                            cached_tokens: None,
                            estimated: false,
                        });
                    }
                }
            }
        }

        if let Some(usage_state) = usage {
            emit(ChatEvent::Usage { usage: usage_state });
        }
        emit(ChatEvent::Done { finish_reason: None });
        Ok(())
    }
}

fn urlencoding_light(input: &str) -> String {
    let mut out = String::new();
    for byte in input.bytes() {
        match byte {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                out.push(byte as char);
            }
            _ => out.push_str(&format!("%{byte:02X}")),
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::providers::anthropic::AnthropicAdapter;

    #[test]
    fn encodes_model_names() {
        assert_eq!(urlencoding_light("gemini-2.0-flash"), "gemini-2.0-flash");
        assert_eq!(urlencoding_light("models/a:b"), "models%2Fa%3Ab");
    }

    #[test]
    fn requires_api_key() {
        assert!(GoogleAdapter::new(String::new(), None).is_err());
        assert!(GoogleAdapter::new("key".to_string(), None).is_ok());
        assert!(AnthropicAdapter::new(String::new(), None).is_err());
        assert!(AnthropicAdapter::new("key".to_string(), None).is_ok());
    }
}
